packages/invoices/src/OutboundShipment/DetailView/reducer.ts

```diff
@@ -249,61 +243,75 @@ export const reducer = (
+            const newLine = {
               ...line,
-              id: lines[existingLineIdx]?.id || line.id,
-              isUpdated: true,
+              isCreated: true,
+              isUpdated: false,
               isDeleted: false,
             };
-          } else {
-            const existingDeletedLineIdx = lines.findIndex(
-              ({ isDeleted, itemId }) => itemId === line.itemId && isDeleted
-            );
-
-            if (existingDeletedLineIdx >= 0) {
-              lines[existingDeletedLineIdx] = {
-                ...lines[existingDeletedLineIdx],
-                ...line,
-                id: lines[existingDeletedLineIdx]?.id || line.id,
-                isCreated: false,
-                isDeleted: false,
-                isUpdated: true,
-              };
-            } else {
-              line.isCreated = true;
-              line.isUpdated = true;
-              line.isDeleted = false;
-              draft.lines.push(createLine(line, draft, dispatch));
-            }
+            const summaryItem = createSummaryItem(line.itemId, [newLine]);
+            items.push(summaryItem);
           }
 
-          draft.update = (key, value) => {
-            dispatch?.(OutboundAction.updateInvoice(key, value));
-          };
-
           break;
         }
 
         case ActionType.DeleteLine: {
           const { draft } = state;
+          const { items } = draft;
           const { payload } = action;
-          const { line } = payload;
 
-          const lineToDelete = draft.lines.find(({ id }) => id === line.id);
+          const { line } = payload;
 
-          if (lineToDelete) {
-            if (lineToDelete.isCreated) {
-              draft.lines = draft.lines.filter(({ id }) => id !== line.id);
+          const { existingRow, existingSummaryItem } = getExistingLine(
+            items,
+            line
+          );
+
+          if (existingRow && existingSummaryItem) {
+            if (existingRow.isCreated) {
+              delete existingSummaryItem.batches[line.id];
+              if (Object.keys(existingSummaryItem.batches).length === 0) {
+                const idx = items.findIndex(
+                  ({ id }) => id === existingSummaryItem.id
+                );
+                items.splice(idx, 1);
+              }
             } else {
-              lineToDelete.isDeleted = true;
+              if (existingSummaryItem.batches[line.id]) {
+                // The if condition above doesn't help TS know that this is guaranteed
+                // to be defined.
+                (
+                  existingSummaryItem.batches[
+                    existingRow.id
+                  ] as OutboundShipmentRow
+                ).isDeleted = true;
+
+                const allDeleted = Object.values(
+                  existingSummaryItem.batches
+                ).every(({ isDeleted }) => isDeleted);
+                if (allDeleted) existingSummaryItem.isDeleted = true;
+              }
```

Comment:
this one was a bit more annoying because we need to check if everythings deleted and mark the summary item or potentially remove it.

Here I wondered how we're going to do deletes for a batch. Possibly put a checkbox in the batch expando?

Reply:
oh! interesting. I don't remember/forgot about that. There is an explicit delete mutation (which seems easier than an update with 0 quantity?)

What I meant was that if we have a table of items grouped by item, from which you use the expand to view the batches, then the current checkbox would be on the summary item row, so would need some alternative for deleting a specific batch

Reply:
ah, yes. that becomes a bit of a mess. force people to edit the row and adjust quantities? Delete the row and re-add it?
I prefer the explicit mutation to an implied delete where you set quantity to 0 and save and the row disappears. or something like that.

Reply:
I've kept the reducer action as per line for now. In the expando in the table, we could just add extra check boxes and adjust the 'table state' to account for 'sub rows'. Then we can have the summary item row checkbox check all the sub rows and delete the whole item, or just check the subrows in the expando for individual batches. Hope that makes sense 😨